lib/taskcluster-config.js

```diff
@@ -0,0 +1,148 @@
+var debug   = require('debug')('github:worker');
+var yaml    = require('yaml-js');
+var Promise = require('promise');
+var slugid  = require('slugid');
+var tc      = require('taskcluster-client');
+var jparam  = require('json-parameterization');
+var _       = require('lodash');
+
+var taskclusterConfig = module.exports = {};
+
+/**
+ * Compares a list of expressions and a list of values,
+ * returning true if any possible combination is a match
+ **/
+function listContainsExpressions(expressions, values) {
+  for (var i in expressions ) {
+    let exp = RegExp(expressions[i], 'i');
+    // just join values so that we don't have to compare multiple times
+    if (exp.test(values.join(' '))) return true;
+  }
+  return false;
+};
+
+function genGitHubEnvs(payload) {
+  return {
+    GITHUB_EVENT: payload.details['event.type'],
+    GITHUB_BRANCH: payload.details['event.base.repo.branch'],
+    GITHUB_PULL_REQUEST: payload.details['event.pullNumber'],
+    GITHUB_BASE_REPO_URL: payload.details['event.base.repo.url'],
+    GITHUB_BASE_USER: payload.details['event.base.user.login'],
+    GITHUB_BASE_SHA: payload.details['event.base.sha'],
+    GITHUB_BASE_BRANCH: payload.details['event.base.repo.branch'],
+    GITHUB_BASE_REF: payload.details['event.base.ref'],
+    GITHUB_HEAD_REPO_URL: payload.details['event.head.repo.url'],
+    GITHUB_HEAD_USER: payload.details['event.head.user.login'],
+    GITHUB_HEAD_SHA: payload.details['event.head.sha'],
+    GITHUB_HEAD_BRANCH: payload.details['event.head.repo.branch'],
+    GITHUB_HEAD_REF: payload.details['event.ref'],
+    GITHUB_HEAD_USER_EMAIL: payload.details['event.head.user.email']
+  };
+};
+
+/**
+ * Attach fields to a compiled taskcluster github config so that
+ * it becomes a complete task graph config.
+ **/
+function completeTaskGraphConfig(taskclusterConfig, payload) {
+  taskclusterConfig.scopes = [
+    'queue:*',
+    'docker-worker:*',
+    'scheduler:*'
+  ];
+
+  taskclusterConfig.routes = [
+    `taskcluster-github.${ payload.organization }.${ payload.repository }.` + payload.details['event.head.sha']
+  ];
+
+  // each task can optionally decide if it wants github specific environment
+  // variables added to it
+  taskclusterConfig.tasks = taskclusterConfig.tasks.map((task) => {
```

Comment:
Okay, I just did object substitution and proposed a more expressive format on:
https://github.com/jonasfj/json-parameterization/pull/1

I know I mixed the discussion a bit... but just started writing at this is where it ended..